ENIGMAsystem/SHELL/Graphics_Systems/General/GSprimitives.cpp

```diff
@@ -1,17 +1,254 @@
+#endif
+
+namespace {
+
+// the batching mode is initialized to the default here
+int draw_batch_mode = enigma_user::batch_flush_deferred;
+// the texture that was specified when the current primitive batch began
+int draw_batch_texture = -1;
+// whether a batch has been started but not flushed yet
+bool draw_batch_dirty = false;
+// lazy create the batch stream that we use for combining primitives
+int draw_get_batch_stream() {
+  static int draw_batch_stream = -1;
+  if (!enigma_user::d3d_model_exists(draw_batch_stream))
+    draw_batch_stream = enigma_user::d3d_model_create(enigma_user::model_stream);
+  return draw_batch_stream;
+}
+// helper function for beginning a deferred batch to determine when texture swap occurs
+void draw_batch_begin_deferred(int texId) {
+  if (draw_batch_texture != texId) {
+    enigma_user::draw_batch_flush(enigma_user::batch_flush_deferred);
+    draw_batch_texture = texId;
+  }
+  draw_batch_dirty = true;
+}
+
+} // anonymous namespace
+
+namespace enigma_user
+{
+
+unsigned draw_primitive_count(int kind, unsigned vertex_count) {
+  switch (kind) {
+    case pr_pointlist: return vertex_count;
+    case pr_linelist: return vertex_count / 2;
+    case pr_linestrip: if (vertex_count > 1) return vertex_count - 1; break;
+    case pr_trianglelist: return vertex_count / 3;
+    case pr_trianglestrip: if (vertex_count > 2) return vertex_count - 2; break;
+    case pr_trianglefan: if (vertex_count > 2) return vertex_count - 2; break;
+    #ifdef DEBUG_MODE
+    default: show_error("Unknown primitive kind: " + std::to_string(kind), true);
+    #endif
+  }
+  return 0;
+}
+
+void draw_batch_flush(int kind) {
+  static bool flushing = false;
+
+  // return if the kind of flush being requested
+  // is not the mode of flushing we have enabled
+  if (draw_batch_mode != kind) return;
+
+  // guard against infinite recursion in case this flush
+  // leads to other state changes that trigger another flush
+  if (flushing || !draw_batch_dirty) return;
+  flushing = true;
+
+  // the never flush mode means the batch isn't drawn
+  // we must still clear it from memory to avoid leaks
+  if (draw_batch_mode != batch_flush_never) {
+    d3d_model_draw(draw_get_batch_stream(), draw_batch_texture);
+  }
+  d3d_model_clear(draw_get_batch_stream());
+
+  flushing = false;
+  draw_batch_dirty = false;
+}
+
+void draw_set_batch_mode(int mode) {
+  if (draw_batch_mode != batch_flush_immediate &&
+      draw_batch_mode != mode) {
+    // if we were not doing immediate flushing before
+    // we need to flush now before we switch modes or
+    // the batch may never be drawn
+    draw_batch_flush(draw_batch_mode);
+  }
+  draw_batch_mode = mode;
+}
+
+int draw_get_batch_mode() {
+  return draw_batch_mode;
+}
+
+void draw_primitive_begin(int kind)
+{
+  draw_batch_begin_deferred(-1);
+  d3d_model_primitive_begin(draw_get_batch_stream(), kind);
+}
+
+void draw_primitive_begin_texture(int kind, int texId)
+{
+  draw_batch_begin_deferred(texId);
```

Comment:
@JoshDreamland The actual batching logic is here, when you begin a primitive. It decides to break the current batch if the texture is different, or else continue combining primitives into the batch model. It does that for deferred mode. In immediate mode it just flushes when the primitive ends.